### PR TITLE
[DataGrid] Add a new demo with sparklines

### PR DESCRIPTION
**Pull request by flaviendelangle**
Doc preview: https://deploy-preview-9228--material-ui-x.netlify.app/x/react-data-grid/custom-columns/

- [x] Allow to entirely hide an axis on the cartesian charts (@alexfauquette did I miss a way to do it? If so maybe we should document it)
- [x] Add a demo (not sure where to put it) on the data grid doc with a sparkline column type

## Challenges

- Performances: The charts should be performant enough with small datasets to not freeze the scroll of the data grid

- Demo: Which dataset should we use? The npm data I used works well, but it's yet another dataset never used elsewhere

### CODE REVIEW

**Comment by flaviendelangle** [comment]
@cherniavskii @alexfauquette @richbustos 

Where do you think this demo should live?

And do you think we should set the column data inside one of the data generators (commodities or employees) or do you think we should hardcode it inside the demo (the demo could still use the data generator for the other columns)?

**Comment by joserodolfofreitas** [comment]
> Where do you think this demo should live?

I believe we need a page called "Custom columns" under the **Columns** menu where we can showcase and instruct users to customize columns with some examples (like the star-rating column); There, we can add the example for the sparklines and date pickers as well. 
And as we use demos with the star-rating column, I believe it's a good idea to add the sparklines to the data generator and display it in other demos around the docs.

Later when we have more integrations with charts, we should aim for a specific menu to cover those.

File: docs/data/data-grid/column-definition/SparklineColumn.tsx [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  const xAxisData = React.useMemo(
+    () => Array.from({ length: data.length }).map((_, index) => index),
+    [data],
+  );
+
+  return (
+    <LineChart
+      margin={{ left: 6, right: 6, top: 4, bottom: 4 }}
+      yAxis={[{ hidden: true }]}
+      xAxis={[{ hidden: true, data: xAxisData }]}
```
Comment (alexfauquette) [inline docs/data/data-grid/column-definition/SparklineColumn.tsx:29]: If I'm correct (a bit lost between what is already committed and what is in open PR)

Your `xAxisData` is already the default `xAxis.data` provided for `LineCharts`

To hide axis, you can use `null` (here is the [docs example](https://mui.com/x/react-charts/axis/#position))

```suggestion
      leftAxis={null}
      bottomAxis={null}
```

**Comment by richbustos** [comment]
> Where do you think this demo should live?

After quickly looking and familiarizing myself with the docs, I agree with @joserodolfofreitas 

Separating out the demo imo will help the users understand the examples being used.

File: docs/data/data-grid/column-definition/SparklineColumn.tsx (hunk 1 above)
Comment (flaviendelangle) [reply docs/data/data-grid/column-definition/SparklineColumn.tsx:29]: It would be nice to add this to the doc, I did not find it myself :+1:

Comment (alexfauquette) [reply docs/data/data-grid/column-definition/SparklineColumn.tsx]: where would you add it?

Comment (alexfauquette) [reply docs/data/data-grid/column-definition/SparklineColumn.tsx]: So splitting the [Position](https://deploy-preview-9228--material-ui-x.netlify.app/x/react-charts/axis/#position) section in two: "hide the axis" and "place the axis"

**Comment by cherniavskii** [comment]
Since the charts package is not published yet, and charts docs are not public, is there any value in making this demo public?
Would it make more sense to move this demo to a separate page under the `Columns` pages group and show it only in deploy previews (as we currently do with Charts docs)?

**Comment by flaviendelangle** [comment]
> Since the charts package is not published yet, and charts docs are not public, is there any value in making this demo public?

For me the idea is to merge this PR the same week we release the alpha (I hope this week) or as soon as possible after.

**Comment by richbustos** [comment]
What's the status of this change?

**Comment by flaviendelangle** [comment]
Done, @cherniavskii @joserodolfofreitas I think it's better if I give you the ownership of this PR to update everything related to the data grid :+1:

**Comment by oliviertassinari** [comment]
> Performances: The charts should be performant enough with small datasets to not freeze the scroll of the data grid

1. We render the mark plot to later hide them with CSS:

Screenshot 2023-08-12 at 17 25 42

This diff significantly improves the performance on my end.

```diff
diff --git a/packages/x-charts/src/LineChart/LineChart.tsx b/packages/x-charts/src/LineChart/LineChart.tsx
index f2377b9ad..274733819 100644
--- a/packages/x-charts/src/LineChart/LineChart.tsx
+++ b/packages/x-charts/src/LineChart/LineChart.tsx
@@ -69,7 +69,6 @@ function LineChart(props: LineChartProps) {
         <LinePlot />
       </g>
       <Axis topAxis={topAxis} leftAxis={leftAxis} rightAxis={rightAxis} bottomAxis={bottomAxis} />
-      <MarkPlot />
       <Highlight {...highlight} />
       <Tooltip {...tooltip} />
       <ClipPath id={clipPathId} />
```

So likely we want to think of a hide API

2. Creating one `styled()` element per MarkElement will be much slower than one per MarkPlot. Even with the move to static CSS extraction, each React component intermediary ads about 10% of performance overhead (what I could measure in the past). So the more we can flatten the React tree, the better.

**Comment by flaviendelangle** [comment]
The comment regarding the mark plot do not apply to the sparkline built by @alexfauquette so no problem on that side

**Comment by romgrk** [comment]
re @oliviertassinari's ideas: I'm not sure I'd spend much time trying to optimize the current implementation. The root of the issue is that react+svg rendering is too slow (by experience, and I've done quite a few charts) and I don't think we'll end up with a usable sparkline anyway (for the datagrid use-case). We would get a better speedup by not using react for rendering the SVG content and instead rendering it imperatively, in the same manner that one would implement canvas rendering.

**Comment by oliviertassinari** [comment]
> 1. the mark plot do not apply to the sparkline built by

@flaviendelangle 👍  Maybe only a customization question, then for line charts, if people don't want the mark.

Point 2. might apply https://mui.com/x/react-charts/sparkline/#basics. It looks like we call styled() too often:

Screenshot 2023-08-14 at 17 55 49

File: docs/data/data-grid/column-definition/column-definition.md [hunk 2]
```diff
@@ -302,26 +302,9 @@ However, some types require additional properties to be set to make them work co
 
 {{"demo": "ColumnTypesGrid.js", "bg": "inline"}}
 
-## Custom column types
+### Custom column types
 
-You can extend the native column types with your own by simply spreading the necessary properties.
```
Comment (cherniavskii) [inline docs/data/data-grid/column-definition/column-definition.md:307]: Moved to https://deploy-preview-9228--material-ui-x.netlify.app/x/react-data-grid/custom-columns/

File: docs/data/data-grid/recipes-editing/recipes-editing.md [hunk 3]
```diff
@@ -105,18 +105,3 @@ Using the [controlled mode](/x/react-data-grid/editing/#controlled-model) and li
 The following demo implements this behavior.
 
 {{"demo": "SingleClickEditing.js", "bg": "inline", "defaultCodeOpen": false}}
-
```
Comment (cherniavskii) [inline docs/data/data-grid/recipes-editing/recipes-editing.md:108]: Moved to https://deploy-preview-9228--material-ui-x.netlify.app/x/react-data-grid/custom-columns/

**Comment by cherniavskii** [comment]
I've expanded the custom columns section and added a pickers demo: https://deploy-preview-9228--material-ui-x.netlify.app/x/react-data-grid/custom-columns/

It is ready for review now.

**Comment by alexfauquette** [comment]
I don't know why but `shape-rendering: crispedges` that I used to avoid tiny white lines between rectangles in usual charts makes some bar disappear 

https://github.com/mui/mui-x/assets/45398769/ef23ff35-9844-4dad-8a55-f65dbebe24c2

Maybe just removing this CSS property from sparkline

**Review by alexfauquette — APPROVED** [review]
Thanks @cherniavskii and @flaviendelangle for taking care of this first integration 🙏

**Review by joserodolfofreitas — APPROVED** [review]
Great work, everyone!

**Pull request merged**